theia-extensions/viewer-prototype/src/browser/trace-viewer/trace-viewer-toolbar-contribution.tsx

```diff
@@ -0,0 +1,106 @@
+import * as React from 'react';
+import { injectable, inject } from 'inversify';
+import { AbstractViewContribution, ApplicationShell } from '@theia/core/lib/browser';
+import { TabBarToolbarContribution, TabBarToolbarRegistry } from '@theia/core/lib/browser/shell/tab-bar-toolbar';
+import { CommandRegistry, DisposableCollection, Emitter, MenuModelRegistry } from '@theia/core';
+import { ContextMenuRenderer } from '@theia/core/lib/browser';
+import { signalManager, Signals } from '@trace-viewer/base/lib/signals/signal-manager';
+import { TraceViewerWidget } from './trace-viewer';
+import { TspClientProvider } from '../tsp-client-provider-impl';
+import { TraceExplorerViewsWidget } from '../trace-explorer/trace-explorer-sub-widgets/trace-explorer-views-widget';
+import { TraceViewerToolbarFilterMenus, TRACE_VIEWER_TOOLBAR_COMMAND_FILTER } from './trace-viewer-toolbar-commands';
+
+@injectable()
+export class TraceViewerToolbarContribution extends AbstractViewContribution<TraceViewerWidget> implements TabBarToolbarContribution {
+    @inject(ApplicationShell) protected readonly shell: ApplicationShell;
+    @inject(ContextMenuRenderer) protected readonly contextMenuRenderer!: ContextMenuRenderer;
+    @inject(TspClientProvider) protected readonly tspClientProvider!: TspClientProvider;
+    @inject(MenuModelRegistry)
+    protected readonly menus: MenuModelRegistry;
+    @inject(CommandRegistry)
+    protected readonly commands: CommandRegistry;
+    private outputAnnotationsMap: Map<string, { categoryName: string, toggleInd: boolean }[]> = new Map<string, { categoryName: string, toggleInd: boolean }[]>();
+    private onAnnotationsFetchedSignal = (payload: { [key: string]: { categoryName: string, toggleInd: boolean }[] }) => this.doHandleAnnotationsFetchedSignal(payload);
+    protected readonly onAnnotationsChangedEmitter = new Emitter<void>();
+    protected readonly onAnnotationsChangedEvent = this.onAnnotationsChangedEmitter.event;
+
+    constructor() {
+        super({
+            widgetId: TraceExplorerViewsWidget.ID,
+            widgetName: TraceExplorerViewsWidget.LABEL,
+            defaultWidgetOptions: {
+                area: 'right',
+            },
+        });
+        signalManager().on(Signals.ANNOTATIONS_FETCHED, this.onAnnotationsFetchedSignal);
+    }
+
+    initializeLayout(): void {
+        this.openView({ activate: false });
+    }
+
+    private doHandleAnnotationsFetchedSignal(payload: { [key: string]: { categoryName: string, toggleInd: boolean }[] }) {
+        Object.entries(payload).forEach(([key, value]) => {
+            if (value.length === 0) {
+                this.outputAnnotationsMap.delete(key);
```

Comment:
Looks like there's only one global list. So when you have two traces opened with the same view output opened, and you remove the output from one trace, the global list is empty and the filter is disabled.

Then when you switch back to the other trace, the filter is still disabled even though that trace still has an output opened.

Reply:
Yeah that is a known issue. If you have multiple tabs open then the logic to add/remove filter categories based on the output id fails.